### ISSUE TITLE
Option from C# Try method

### ISSUE THREAD

**Issue by 3Rafal** [fsprojects/FSharpPlus#408]
When working with C# APIs there are cases when I have an interface

```csharp
public interface IValueGetter
{
    bool TryGetValue(out var myValue);
}
```
So I map it to option like

```fsharp
let optionFromTryGet = function
     | (true, x) -> Some x
     | _ -> None

let optValue = getter.TryGetValue() |> optionFromTryGet
```
I wonder if it can be generalized to Option extension in F#+.
Something like `Option.fromTry` implemented as above helper function.

There are similar implementations already floating around on the internet.
For example this one, which is dedicated to parsing: 
http://www.fssnip.net/2y/title/Functional-wrappers-for-TryParse-APIs

**Comment by wallymathieu** [comment]
See 
https://github.com/fsprojects/FSharpPlus/blob/master/tests/FSharpPlus.Tests/General.fs#L2121-L2136

where you have the customer id from: https://github.com/fsprojects/FSharpPlus/blob/master/tests/CSharpLib/CustomerId.cs#L5

**Comment by 3Rafal** [comment]
@wallymathieu , 
I would like to generalize over C# try pattern with out variable. It's used not only for parsing, but also on getting values in general.
See Dictionary example: https://docs.microsoft.com/pl-pl/dotnet/api/system.collections.generic.dictionary-2.trygetvalue?view=net-5.0

This kind of methods return tuple of `bool * 'a` and I would like to be able to convert it to option easily. I've already written this kind of handling a couple of times in my projects, so I guess it could be generalized to a pattern on F# side.

**Comment by 3Rafal** [comment]
Also, this pattern was described by Scott Wlaschin here:
https://swlaschin.gitbooks.io/fsharpforfunandprofit/content/posts/completeness-seamless-dotnet-interop.html

He uses couple of different methods with out variable and binds them to `bool * 'a` tuples. But, he doesn't show that you have to pattern match on these tuples to use them. What I would like to do is to pipe them directly to Option monad, so that the handling is easier.

**Comment by wallymathieu** [comment]
I'm unsure if it's possible, but it would be nice to have.

**Comment by gusty** [comment]
Isn't that exactly what our `Option.protect` function does? 

http://fsprojects.github.io/FSharpPlus/reference/fsharpplus-option.html#protect

when combined with an `Option.ofPair` function we could possible add?

**Comment by 3Rafal** [comment]
@gusty , I don't think it's connected with `protect`. Protect works on Exceptions mechanism, and try with out works on two return values (one is a bool returned directly, another is a value returned via out parameter).

I've added implementation in #409. I think unit tests show how I think about these use cases.

**Comment by wallymathieu** [comment]
Option.ofPair looks nice 👍 , but the name might need some thought.

**Comment by 3Rafal** [comment]
It was suggested by @gusty, so I made it this way. I think it's neat and terse, but I can change it if you want to.

**Comment by wallymathieu** [comment]
I'm unsure. In one way it makes sense in that it's a pair. It describes a specific pattern, though the name for it does help with naming the function. Perhaps the ofPair is good enough with a bit of docs?

**Comment by 3Rafal** [comment]
I've mentioned this in docstrings. I wonder if it's ok, or I should expand it?

**Comment by 3Rafal** [comment]
@gusty , I guess we can close it :-)

**Comment by gusty** [comment]
Sure, closed by #409
